### DISCUSSION TITLE
Filtering by Hybrid Property not working

### DISCUSSION THREAD

**Discussion by VelkovFilip** [sqlalchemy/sqlalchemy#5904] [Usage Questions] [answered]
The problem occurs when I am trying to filter the results depending on a boolean hybrid field. The model is as follows:

```py

class Restaurant(Base):
    __tablename__ = 'Restaurant'
    id = Column(Integer, primary_key = True)
    open_from = Column(Time)
    open_to = Column(Time)
    @hybrid_property
    def before_midnight(self):
        return self.open_from <= self.open_to
    @hybrid_property
    def is_Open(self):
        time = datetime.datetime.now().time()
        if self.before_midnight == True:
            return (self.open_from <= time) & (time <= self.open_to)
        else:
            return  (time<self.open_to) | (time>self.open_from)

```
To test it I am creating 2 instances of this class:

```py
import datetime

fromm1 = datetime.time(10,30)
to1 = datetime.time(00,30)

fromm = datetime.time(6,30)
to = datetime.time(9,30)

res = Restaurant(id = 1,open_from = fromm1,open_to = to1)
res2 = Restaurant(id = 2, open_from = fromm, open_to = to)

```

After adding them to the database I am querying the database for these objects and I'm storing them in the variables *r1* and *r2* respectively.  

Now if I check for the hybrid fields' values manually I get the following result:
```py
r1.is_Open,r2.is_Open

Out[]: (True, False)
```

However if I try to filter the instances by this field I get different results:

```py
a = session.query(Restaurant).filter(Restaurant.is_Open == True).all()
a[0].id,a[1].id

Out[]: (1,2)
```

You may find the complete code [here](https://gist.github.com/VelkovFilip/314df2d08042a6f2863ebe944a8e357d)

**Comment by zzzeek** [comment, marked as answer]
hi there -

the purpose of @hybrid_property is to provide *two* implementations of a function.  One implementation runs in **Python, specific to an instance of your model**, and the other implementation **generates a SQL expression, specific to the table to which your model is mapped against**.

In your example, "is_open" looks like this:

```
   @hybrid_property
    def is_Open(self):
        time = datetime.datetime.now().time()
        if self.before_midnight == True:
            return (self.open_from <= time) & (time <= self.open_to)
        else:
            return  (time<self.open_to) | (time>self.open_from)
```

in this function, we can identify the following element that isn't appropriate for SQL:

1. `"if self.before_midnight  ==  True"` - when all you have is a `Restaurant` class, `Restaurant.before_midnight` is neither Python ``True`` nor ``False`` - it's the SQL expression "restaurant.open_from <= restaurant.open_to"

in SQL, the equivalent of the "if" statement in a SQL expression is the [CASE](https://www.w3schools.com/sql/sql_case.asp) expression which SQLAlchemy provides as https://docs.sqlalchemy.org/en/13/core/sqlelement.html?highlight=case#sqlalchemy.sql.expression.case.   So you would want to add a new method to your class that uses `@expression` and is along the lines of:


```
@is_Open.expression
def is_Open(cls):
    # this could also be the Python datetime.now().time()
    time = func.current_time()
    return case(
        [
            (cls.before_midnight, (cls.open_from <= time) & (time <= self.open_to))
        ],
        else_=(time < cls.open_to) | (time>cls.open_from)
    )
```


note that I also used the SQL time() function (this may vary depending on what database you are targeting). that would help in the case that a SQL expression you've created is being re-used many times, so that it always uses the correct time (assuming your database server time is correct) even if the SQL expression has been present in the application's memory for a long time.

however with both datetime.now().time() and if you wanted to use a SQL time function, you probably also want to take timezones into account, as getting now().time() gives you a number that's not very meaningful if the user of your application is in a different timezone than where the program is running.

**Reply by VelkovFilip** [reply]
Thanks a lot @zzzeek!
